File: gesture-controller.js
class GestureController {
    constructor() {
        this.hands = new Hands({
            locateFile: (file) => `https://cdn.jsdelivr.net/npm/@mediapipe/hands/${file}`
        });
        this.faceMesh = new FaceMesh({
            locateFile: (file) => `https://cdn.jsdelivr.net/npm/@mediapipe/face_mesh/${file}`
        });
        
        this.videoElement = document.getElementById('gesturePreview');
        this.lastFaceTime = Date.now();
        this.isPaused = false;
        
        this.setupHandlers();
    }

    setupHandlers() {
        // Configure MediaPipe instances
        this.hands.setOptions({
            maxNumHands: 1,
            modelComplexity: 1,
            minDetectionConfidence: 0.5,
            minTrackingConfidence: 0.5
        });

        this.faceMesh.setOptions({
            maxNumFaces: 1,
            minDetectionConfidence: 0.5,
            minTrackingConfidence: 0.5
        });

        // Set up result handlers
        this.hands.onResults(this.onHandResults.bind(this));
        this.faceMesh.onResults(this.onFaceResults.bind(this));
    }

    async start() {
        // Create camera feed
        const stream = await navigator.mediaDevices.getUserMedia({ video: true });
        this.videoElement.srcObject = stream;
        this.videoElement.play();

        // Start detection loops
        this.startDetectionLoop();
    }

    startDetectionLoop() {
        if (!this.videoElement) return;

        const detectFrame = async () => {
            if (this.videoElement.videoWidth > 0) {
                // Process frame through both detectors
                await this.hands.send({ image: this.videoElement });
                await this.faceMesh.send({ image: this.videoElement });
            }
            // Request next frame
            requestAnimationFrame(detectFrame);
        };

        detectFrame();
    }

    async onHandResults(results) {
        if (results.multiHandLandmarks && results.multiHandLandmarks.length > 0) {
            const landmarks = results.multiHandLandmarks[0];
            const gestureData = {
                frame: this.videoElement,
                landmarks: landmarks
            };

            // Send gesture data to backend
            try {
                const response = await fetch('/api/gesture/state', {
                    method: 'POST',
                    headers: {
                        'Content-Type': 'application/json'
                    },
                    body: JSON.stringify(gestureData)
                });
                
                const result = await response.json();
                this.handleGestureAction(result);
            } catch (error) {
                console.error('Error sending gesture data:', error);
            }
        }
    }

    async onFaceResults(results) {
        if (results.multiFaceLandmarks && results.multiFaceLandmarks.length > 0) {
            this.lastFaceTime = Date.now();
            const landmarks = results.multiFaceLandmarks[0];
            const faceData = {
                frame: this.videoElement,
                landmarks: landmarks
            };

            try {
                const response = await fetch('/api/gesture/state', {
                    method: 'POST',
                    headers: {
                        'Content-Type': 'application/json'
                    },
                    body: JSON.stringify(faceData)
                });
                
                const result = await response.json();
                this.handleGestureAction(result);
            } catch (error) {
                console.error('Error sending face data:', error);
            }
        } else if (Date.now() - this.lastFaceTime > 2000 && !this.isPaused) {
            // Auto-pause if face not detected for 2 seconds
            this.handleGestureAction({ action: 'pause', value: null });
        }
    }

    handleGestureAction(actionData) {
        const videoPlayer = document.querySelector('.video-player');
        if (!videoPlayer) return;

        switch (actionData.action) {
            case 'volume_up':
                videoPlayer.volume = Math.min(1, videoPlayer.volume + 0.1);
                break;
            case 'volume_down':
                videoPlayer.volume = Math.max(0, videoPlayer.volume - 0.1);
                break;
            case 'seek_forward':
                videoPlayer.currentTime += actionData.value;
                break;
            case 'seek_backward':
                videoPlayer.currentTime -= actionData.value;
                break;
            case 'toggle_play':
                if (videoPlayer.paused) {
                    videoPlayer.play();
                    this.isPaused = false;
                } else {
                    videoPlayer.pause();
                    this.isPaused = true;
                }
                break;
            case 'brightness_adjust':
                this.adjustBrightness(actionData.value);
                break;
        }
    }

    adjustBrightness(value) {
        const videoPlayer = document.querySelector('.video-player');
        if (!videoPlayer) return;

        // Apply brightness filter
        const brightness = Math.max(0.5, Math.min(1.5, 1 + value));
        videoPlayer.style.filter = `brightness(${brightness})`;
    }
}

// Initialize controller when requested
function initGestureController() {
    const controller = new GestureController();
    controller.start().catch(console.error);
    return controller;
} 

File: gesture_controller.py
import cv2
import mediapipe as mp
import numpy as np
import pyautogui
import time

class GestureController:
    def __init__(self):
        self.mp_hands = mp.solutions.hands
        self.mp_face_mesh = mp.solutions.face_mesh
        self.hands = self.mp_hands.Hands(max_num_hands=1)
        self.face_mesh = self.mp_face_mesh.FaceMesh(
            min_detection_confidence=0.5,
            min_tracking_confidence=0.5
        )
        self.drawing = mp.solutions.drawing_utils
        self.last_face_time = time.time()
        self.paused = False

    def count_fingers(self, landmarks):
        """Count raised fingers in hand landmarks"""
        cnt = 0
        thresh = (landmarks.landmark[0].y*100 - landmarks.landmark[9].y*100)/2

        if (landmarks.landmark[5].y*100 - landmarks.landmark[8].y*100) > thresh:
            cnt += 1
        if (landmarks.landmark[9].y*100 - landmarks.landmark[12].y*100) > thresh:
            cnt += 1
        if (landmarks.landmark[13].y*100 - landmarks.landmark[16].y*100) > thresh:
            cnt += 1
        if (landmarks.landmark[17].y*100 - landmarks.landmark[20].y*100) > thresh:
            cnt += 1
        if (landmarks.landmark[5].x*100 - landmarks.landmark[4].x*100) > 6:
            cnt += 1

        return cnt

    def process_face(self, frame):
        """Process face detection and head pose"""
        rgb_frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
        results = self.face_mesh.process(rgb_frame)
        face_detected = False

        if results.multi_face_landmarks:
            face_detected = True
            self.last_face_time = time.time()
            
            for face_landmarks in results.multi_face_landmarks:
                # Calculate eye centers for head tilt
                left_eye = np.array([
                    (face_landmarks.landmark[i].x, face_landmarks.landmark[i].y) 
                    for i in [33, 133, 159, 145]
                ])
                right_eye = np.array([
                    (face_landmarks.landmark[i].x, face_landmarks.landmark[i].y) 
                    for i in [362, 263, 386, 374]
                ])
                
                eye_center_x = (left_eye[:, 0].mean() + right_eye[:, 0].mean()) / 2
                face_center_x = face_landmarks.landmark[1].x
                
                # Head tilt controls
                if abs(eye_center_x - face_center_x) > 0.05:
                    return {"action": "brightness_adjust", "value": eye_center_x - face_center_x}

        # Auto-pause if face not detected for 2 seconds
        if not face_detected and time.time() - self.last_face_time > 2:
            return {"action": "pause", "value": None}

        return {"action": "none", "value": None}

    def process_hands(self, frame):
        """Process hand gestures"""
        rgb_frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
        results = self.hands.process(rgb_frame)
        
        if results.multi_hand_landmarks:
            hand_landmarks = results.multi_hand_landmarks[0]
            finger_count = self.count_fingers(hand_landmarks)
            
            # Map finger counts to actions
            actions = {
                1: {"action": "volume_up", "value": 10},
                2: {"action": "volume_down", "value": 10},
                3: {"action": "seek_forward", "value": 5},
                4: {"action": "seek_backward", "value": 5},
                5: {"action": "toggle_play", "value": None}
            }
            
            return actions.get(finger_count, {"action": "none", "value": None})
            
        return {"action": "none", "value": None}

    def process_gesture(self, frame_data):
        """Main gesture processing pipeline"""
        # Convert frame data to numpy array
        frame = np.frombuffer(frame_data, dtype=np.uint8)
        frame = cv2.imdecode(frame, cv2.IMREAD_COLOR)
        
        # Process both face and hand gestures
        face_result = self.process_face(frame)
        hand_result = self.process_hands(frame)
        
        # Prioritize hand gestures over face gestures
        if hand_result["action"] != "none":
            return hand_result
        return face_result 